### DISCUSSION TITLE
layer conceptual identity

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#58] [General]
Kind of reflecting on a lot of things already said and now Hadley's response to @yjunechoe question on layer_* not being a prefix used in gggplot2

> geoms and stats are so tightly intertwined like by like it's not 100% true but by and large like there's a one-to-one correspondence between the geoms and the stats. Like sure you can use stat bin with things other than geom_histogram but kind of mapping the aesthetics in the right way makes it makes it fiddily. So that -- that idea that like combined

I think this logic is probably a little less applicable to more primitive Geoms and their user-facing analogue.  Like, I do want to combine Stat transormations and use them with GeomPoint, or GeomSegment etc. 

But I do think from a user point of view, too, there's like a conceptual identity to the user-facing geom_* and stat_s so that it does  make it worthwhile to fully write out pre-defined layer and not expect a lot of mixing and match on the fly. Like human languages that create new words by smashing combos together are instructive (like German).  Those combo words that are being used aren't being created on the fly - they've already been worked out -- much more organically than a developer -- and people use the existing, pre-defined combos.    At some point, used examples daumenkino -> thumb movie -> flipbook; handschuh -> hand shoe -> glove, in about new geom_*s via Stats talk. Just fun!

---

A downside of the uniqueness of layer identity or conceptual 'oneness' of the layers (in the geom+stat+position sense), I think, means that a lot of the time stat_* and definitely Stats proto objects are kind of ignored from what try to be pretty comprehensive introductions of ggplot2, like my 2020 grammar primer! :-)

![image](https://github.com/user-attachments/assets/88cdc0c8-a0a6-4cf1-ba41-8e5d5f1d2901)

Another example from @andrewheiss, and @-ing him since he's done so much thoughtful ggplot2 and dataviz pedegogy work -- would definitely be interested in any of your thoughts!  

![image](https://github.com/user-attachments/assets/60d18439-5b40-4ef1-9bef-7ac8803cd856)

**Comment by EvaMaeRey** [comment]
Just to keep going about linguistic challenges we face in communicating about ggplot2 and extension, at some point layer started being used to talk about plot components.  So this visualization, and versions of it got really popular and I think it is rather insightful, but elements look like sheets/layers, so we all started calling the graphical components layers.   

![image](https://github.com/user-attachments/assets/2e34721c-1bb9-4908-b9b0-b1c8b8b09550)

**Reply by yjunechoe** [reply]
I used to collect these diagrams a while back (argh - forgot where I kept them!) and I found it really fun to see how different users' mental model of the ggplot grammar translate into this "layered grammar" representation. Partly because users are taught the "declarative" grammar, but these diagrams can be unclear whether they represent the declarative order (the sensible sequence of `+` in user-facing code) or the procedural order (the sequence of operations in the internals to materialize a plot)

For example, the one @teunbrand recently made for ggplot docs is, i think, declarative and doesn't distinguish between stat and geom (the division probably hurts more than it helps for novices):

![image](https://github.com/user-attachments/assets/75d3f4f8-2a50-4a9a-a4de-a84f3136972d)

Whereas I think the one that Thomas used originally had a more procedural flavor. This grammar diagram from his youtube workshop is the more popular one:

![image](https://github.com/user-attachments/assets/0b69f4ee-f181-4bf9-bd16-7f9bea6a5de2)

And I think some people misunderstand this as a recommendation for how to `+` their layers, but that begets the fact that this representation is derived from a different diagram earlier in the slides, which lays out the plot-building process in procedural terms.

![image](https://github.com/user-attachments/assets/47394092-2180-43de-8221-b6fdbbced2ba)

Also, kind of a tangent but, I always wondered whether it would help or hurt the mental model of more advanced users if we popularize a distinction between *positional* and *non-positional* scales. Was reminded of this in a recent Mastodon thread (that we were in together) - https://fosstodon.org/@yjunechoe/113425235512384597

**Reply by EvaMaeRey** [reply]
I haven't seen the last one.  Interesting to see that sheets visual analogy is still used, even as the stack isn't.   Nice collection!

> how to + their layers

Wait, do you refer to coordinates/scales/theme as layers?  

I feel like this is not original documented meaning, in "'layered' grammar of graphics".   Just evolved to also have this additional meaning, cause the schematic is useful, and ggplot2 is so easy to use and 'layer' also feels like a good description of what the graphical poem looks like.

**Reply by yjunechoe** [reply]
> Wait, do you refer to coordinates/scales/theme as layers? I feel like this is not original documented meaning, in "'layered' grammar of graphics". Just evolved to also have this additional meaning

Oh for sure (edit: "for sure" as in "I acknowledge this unfortunate fact" not "I agree with this and do it myself" 😅) - *I* don't think like this in my mental model; just using the average user's terminology here. I think most users are pretty comfortable saying, e.g., a "theme layer", though in the "actual" grammar, Layer is reserved to specifically mean Elements (Elements being Wilkinson's term for it).

**Reply by teunbrand** [reply]
I also think using the word 'layer' for coordinates/scales/themes/facets etc is prone to misinterpretation. 'Layers' implies you can stack an arbitrary number of them (which is true from the geom/stat interpretation of layers), whereas you cannot use 2 facet components or 2 coord components in a plot. Also, even though it feels like you're 'adding' these components to a plot, what you're really doing is replacing/updating the previous `coord_cartesian()`, `facet_null()`, `theme_gray()` default components.

**Reply by yjunechoe** [reply]
I think we're all in agreement here!

I think the confusion of the average user is unfortunate but understandable. For example, the ggplot book uses "layer" a bit loosely (or perhaps, tries to draw a somewhat subtle distinction between "layers" and "the layered grammar") - I wonder how we could make things clearer:

> All together, the **layered grammar** defines a plot as the combination of:
> 
> - A default dataset and set of mappings from variables to aesthetics.
> 
> - **One or more layers**, each composed of a geometric object, a statistical transformation, a position adjustment, and optionally, a dataset and aesthetic mappings.
> 
> - One scale for each aesthetic mapping.
> 
> - A coordinate system.
> 
> - The faceting specification.

https://ggplot2-book.org/mastery.html#sec-components

**Reply by EvaMaeRey** [reply]
Back here after the survey on 'number of layers'!! https://fosstodon.org/@EvaMaeRey/114507934980942568

It seems like the term 'layer/layered' is motivated use to distinguish the ggplot2 implementation from Wilkinson's conception, which allows data and mapping to be locally defined.

![image](https://github.com/user-attachments/assets/6d004d47-9ffc-4cd1-b97b-3a368a1ead6e)

I do think the quoted text about introduces ambiguity, but...

I don't think 'All together, the layered grammar defines a plot as the combination of ...' would be rewritten for clarification 'All together, ggplot2 defines a plot as the combination of *layers*:'  but could be rewritten 'All together, ggplot2 defines a plot as the combination of *components*:'  

> All together, the layered grammar defines a plot as the combination of:

> - A default dataset and set of mappings from variables to aesthetics.

> - One or more layers, each composed of a geometric object, a statistical transformation, a position adjustment, and optionally, a dataset and aesthetic mappings.

> - One scale for each aesthetic mapping.

> - A coordinate system.

> - The faceting specification.

**Reply by EvaMaeRey** [reply]
I think I've maybe used 'element' interchangably w/ 'component' but will try to steer clear of that given Wilkinson's usage!

**Comment by yjunechoe** [comment]
I was hoping you'd start this discussion so I can swoop in and drop my half-baked thoughts here! 😆

> A downside of the uniqueness of layer identity or conceptual 'oneness' of the layers

I feel this and I think I have a much stronger take here (than just that it demotes the relevance of `stat_`/`Stat`) which is: monolithic representations of Layers replicate the problems of chart taxonomies (the one that the gg grammar sought out to solve) at the layer level. So people learn to express layers like the boxplot layer with `geom_boxplot()` and learn how to do that well, but yet feel so utterly helpless when confronted by what should be a very similar and familiar problem like "label the boxplot variables" -- users are left at the mercy of (extension) devs to export a layer like `geom_fivenum()` to get back to drawing what they want.

> by and large like there's a one-to-one correspondence between the geoms and the stats.

Hadley is right to note this and I have to concede this point for ggplot as it currently exists. But I feel like if ggplot formalized the delayed aesthetic evaluation system much earlier, we would have actually ended up with more modular Stats and Geoms. Ex: `StatBoxplot` (perhaps named `StatFivenum` in this counterfactual) would return `min`, `median`, `max`, etc. instead of `ymin`, `y`, `ymax`, and it would be paired up with `GeomBoxplot` (perhaps named `GeomBoxWhisker` in this counterfactual) in a user-facing function `layer_boxplot()` through a careful `after_stat()` default spec such as a mapping of `median`->`y`.

I've always wanted to expand more on this idea (of what I called "sublayer modularity" in my [JSM talk](https://www.youtube.com/watch?v=613Q0j6Kjm0)) but alas, my phd is not in graphics 😅

**Reply by EvaMaeRey** [reply]
> I was hoping you'd start this discussion so I can swoop in ...

At your service! I think we keep having versions of this discussion - but it doesn't get old!  I should use 'sublayer modularity' to talk about this.  

>  users are left at the mercy of (extension) devs to export a layer like geom_fivenum() to get back to drawing what they want.
 
From a user perspective, it feels like some digging/sleuthing needs to happen before knowing how Stats can be used beyond their 'base case', because users don't look at the compute methods and in-between dataframes. ggtrace offers a good way to look at those, or ggplot2::layer_data a more imprecise way.  I wonder about a once per session message approach to stat_*s that have a lot of sublayer modularity possibilities (stuffed Stats?).  Like 'here are some nice vars that we create via the Stat compute and they can be used with after_stat: count, prop, and good things will happen', kind of along the lines of required_aes, optional_aes, have after_stat_aes?  Does that make any sense?

**Reply by EvaMaeRey** [reply]
or maybe 'after_stat_vars'...

**Reply by teunbrand** [reply]
They're called computed variables in the docs. They have no formal representation in the Stat classes, contrary to required/optional aesthetics. For this reason it also hard to automate the documentation of such parameters and all the computed variables are 'artisinally' documented.

**Reply by EvaMaeRey** [reply]
I wonder if geom and stat were declared globally and independently, the everyone might be a little more conversant w/ these ideas. 

like layer_identity() could default to point and no compute on the data, but then you could + on geom_tile() and stat_means() or and then see the evolution to a tiles  and then a single tile at means of x and y.  So geom and stat would be externally changeable.  I think on balance it might be harder to use, but could mean for greater immediate practical understanding about geom+stat combinations.

**Reply by teunbrand** [reply]
I think this would lead to some operator precedence shenanigans. How do you know if a layer modifier belongs to the layer before or thereafter?

**Reply by yjunechoe** [reply]
> geom and stat would be externally changeable

This is exactly the philosophy behind [`{ggbuilder}`](https://mjskay.github.io/ggbuilder/)! I think it makes a convincing case of as a proof of concept (maybe we could bug Matthew Kay to present on this 😆)

> operator precedence shenanigans.

I *think* this is resolved if you use the base-pipe, since that's syntactic sugar resolved first at the level of the parser (before `+`):

```r
rlang::expr(
  ggplot(mtcars, aes(hp, mpg)) +
    geom_point() |> I() +
    geom_smooth()
)
#> ggplot(mtcars, aes(hp, mpg)) + I(geom_point()) + geom_smooth()
```

[`{ggblend}`](https://mjskay.github.io/ggblend/) (again from Matthew) comes to mind as doing this well. I suspect [`{ggfx}`](https://ggfx.data-imaginist.com/) would also have no problems with ambiguity if you use the base pipe

**Reply by teunbrand** [reply]
We should definitely bug Matthew Kay to present some of his work 😅 
I'm now also reminded that ggblend has this `*` syntax to apply a blend to a layer.

**Reply by yjunechoe** [reply]
Reminded by `*`, a fun tangent: Julia's [`AlgebraOfGraphics`](https://aog.makie.org/stable/) also has the ggbuilder-style syntax to some extent, so their equivalent of `geom_smooth() + geom_point()` is:

```julia
penguin_bill *
  linear() *
  mapping(color = :species) +
penguin_bill *
  mapping(color = :species)
```

![image](https://github.com/user-attachments/assets/f1e1061b-0f74-4b73-9f76-2e3d1df41fdd)

Where they reserve `*` for layer-internal composition and `+` for across-layer composition.

In a true algebraic fashion, you can also "factor out" the shared components:

```julia
penguin_bill * (linear() + mapping()) * mapping(color = :species)
```

**Reply by EvaMaeRey** [reply]
> How do you know if a layer modifier belongs to the layer before or thereafter?

I think my ideal would be applies globally, unless locally specified which would mean ignoring global specification.  So same as with aesthetic mapping declarations + aes(color = species)  - applies across all layers where color is relevant. 

That Julia code is so interesting!

**Reply by EvaMaeRey** [reply]
> We should definitely bug Matthew Kay to present some of his work 😅 I'm now also reminded that ggblend has this `*` syntax to apply a blend to a layer.

@mjskay certainly would be amazing to hear from!!  We meet virtually about once a month.  Speakers talk through project motivations and decisions, for about 1/2 hour and then time for discussion.   Please let us know if you have any interest in joining at some point!

**Reply by mjskay** [reply]
Oh this sounds fun!!! I would love to join the meetings and also give a talk sometime. Apologies, I think someone reached out to me awhile back (probably @teunbrand?) and asked, I had intended to get back to them but got sidetracked by day job stuff ... :(

Re: this conversation: yeah, {ggblend} and {ggbuilder} both represent some thoughts I've had on treating layer composition more systematically. 

{ggblend} started with the simple idea to support blend modes, but turned into a layer algebra inspired by the way that Wilkinson often talked about the grammar of graphics as an algebra. So it has a notion of an identity transformation (that simply leaves a layer as-is if multiplied by it) and applies rules of abstract algebra to create transformations of layers. To me the coolest part of this was that after I set up the algebra I quickly realized I could implement a way of adding a "highlight" around a line geometry simply by multiplying the layer by the sum of an adjustment operation and the identity operation:

```r
geom_line() * (adjust(linewidth= 2, color = "white") + 1)
```

Which works because `geom_line() * adjust(linewidth = 2, color = "white")` is equivalent to `geom_line(linewidth = 2, color = "white")` and `geom_line() * 1` is just `geom_line()`

{ggbuilder} is more on the side of trying to decompose a layer into a data transformation pipeline, partly because I don't like `after_stat` / `after_scale` / etc, and partly inspired by a talk @yjunechoe gave. The part of that experiment that I never finished, which I would still like to do, is to make it possible to chain multiple `stats` together before feeding them into a single `geom`. I have had several applications where this would be useful, but I never fully figured out how to get it to work.

I also don't know if there's a principled way to combine the layer algebra ideas in {ggblend} with the pipeline idea in {ggbuilder}.

**Reply by EvaMaeRey** [reply]
Early 2025?!? https://teunbrand.github.io/ggplot-extension-club/meetings.html  Would be really interesting to hear from you on these projects!

**Reply by mjskay** [reply]
Yeah early 2025 is doable; simplest times for me are probably monday or friday late morning through evening central time. Tues-Thurs is also doable depending on the week.

I'm happy to talk about the ideas above about layer algebras and/or layer pipelines. Would also be happy to talk about [ggdist](https://mjskay.github.io/ggdist/) and the ideas in there (mostly about bringing probabilistic programming notation to the grammar of graphics). From the implementation side, ggdist also has some (I think) cool ideas for reducing ggplot2 boilerplate and generating comprehensive docs on a lot of stats/geoms using meta-programming.

That's probably too much stuff for one talk, so I'd be happy to focus on whatever people think would be interesting.

**Reply by mjskay** [reply]
(or honestly, I could just throw all that stuff into a talk and see where the discussion goes --- I dunno the length / format of these meetings)

**Reply by EvaMaeRey** [reply]
Typically half hour talk and then time for questions.  Maybe we could schedule slightly longer for you since you've been so active and there might be even more a lot of gg theoretical as well as practical ground to cover.  

Thanks for your schedule details.  I'm going to pencil you in for early 2025!  Will be in touch to narrow to day/time!  🎉🎊🙌

**Reply by EvaMaeRey** [reply]
Or we could schedule you for multiple sessions on the different topics.   Just rereading, and it's a lot!

**Reply by mjskay** [reply]
Happy to do a longer session or multiple sessions, whatever makes sense! I love to talk about this stuff but also don't want to monopolize time.

**Reply by teunbrand** [reply]
I'd be happy to hear about whatever ggplot2 adjacent project you're excited about. We interacted over github a few times but it is always nice when people explain their projects!

**Reply by EvaMaeRey** [reply]
@mjskay are you still interested in presenting in the new year? 🎉

**Reply by mjskay** [reply]
@EvaMaeRey yeah! sorry, busy start to the year. Things calmer now, I'm definitely interested. Let me know what times might be good!

**Reply by EvaMaeRey** [reply]
Sounds great!  I'll reach out after Teun's talk Tuesday to schedule.

**Comment by teunbrand** [comment]
I'm the wrong person to ask for this kind of stuff. I know reasonably well how to express myself in English, but my main grammar/syntax instruction wasn't in English, so I have no idea what 'declarative' and 'interrogative' and 'moods' are and how they relate to language. I can barely tell apart what 'nouns' and 'verbs' are in English while I have no trouble finding the 'workwords' and 'independent namewords' of a sentence.

**Reply by yjunechoe** [reply]
TBH I also know less about what those labels mean for language -- categories like a "declarative sentence" is a grammar-school construct which is too loose to be meaningful for theoretical linguistics!

But for "declarative" vs. "procedural" specifically, I think I actually picked up that lingo from how people talk about ggplot2 and the grammar of graphics. IIUC these map onto specific concepts in programming paradigms ([declarative](https://en.wikipedia.org/wiki/Declarative_programming) vs. [procedural](https://en.wikipedia.org/wiki/Procedural_programming)), beyond just metaphorical relations to language.

**Reply by teunbrand** [reply]
>  grammar-school construct which is too loose to be meaningful for theoretical linguistics!

June, your inner syntactician is hanging out!

I think in terms of that distinction, I guess ggplot2 is mostly declaritive, no? As a user, you needn't bother (most of the time) about the order in which ggplot2 handles plot building (the procedural bit). This is in contrast to base plotting, where probably procedure matters more (e.g. you loop over data groups and jot down something on the graphics device). You need to know e.g. the x limits beforehand if you will be building a plot with >1 graphical layer.

**Reply by EvaMaeRey** [reply]
Analogies break down, but I think the english grammar - even though imperfect -- helped organize that project and keep me motivated.  

Workwords sound like verbs. independent namewords sounds like 'subject' of a sentence.  Fun!

Defer to @yjunechoe on any deep and evidenced linguistic instights.  And greater success of folks using layers versus engaging with sublayers might be the product of just having a lot of complete layers around v. having to compose them.  W/ practice, sublayer use would flow much more.

**Reply by teunbrand** [reply]
> Workwords sound like verbs. independent namewords sounds like 'subject' of a sentence. Fun!

Spot on with verbs! Unfortunately 'subject' translates 1:1 so can't pull any weird Dunglish tricks here :( The independent namewords are nouns

>  sublayer use would flow much more.

I think as soon as people learn how to use stats, it becomes more important and the mark of somebody who understands ggplot2 well. Before I knew how to use stats, I was always just wrangling data myself before handing the data off to some graphical primitive geom.

**Reply by EvaMaeRey** [reply]
and dependent namewords? pronouns?

**Reply by teunbrand** [reply]
The dependent namewords are possessive as in 'my', 'your', 'their' and are always followed by the independent namewords like 'bike', 'house', 'car'. Pronouns are 'front namewords' or 'before namewords', which kinda makes sense with the 'pro'-prefix.

**Reply by yjunechoe** [reply]
Wow I swear this is my first time hearing "workword" and "(independent/dependent) nameword"! 😮

Also, I recognize I'm very blatantly an outlier so please don't mind my comments on analogies to language - I've just been very brainwashed by my program to think about language differently 😆

---

P.S. funny you should mention "subject" - that's a very classic problem in syntax! I think most non-linguist would be happy thinking of a subject as "an agent/experiencer of some action/event" but in Chomsky-an [generative transformational grammar](https://en.wikipedia.org/wiki/Transformational_grammar), such "part of speech" properties emerge not from the meaning of words, but from their syntactic distribution and their internal derivation in an abstract, tree-like representation.

So instead of "a subject is an agent of an action", we say "_Subject_ refers to a syntactically privileged case-assigning position in Spec-IP that can be occupied by a nominal. The subject nominal originates VP-internally as an argument of the verb and is assigned its theta role there, before moving to Spec-IP to receive case."

In case you're curious what that even means (and what a linguist, specifically a syntactician, even does), (1) below is a "syntax tree" that I had to draw in my first year as a proof of this theorem (called "subject raising"). The scope ambiguity in the sentence "Kim does not sell state secrets, as the senators stated" as to what the senators stated (that Kim does? or does not? sell state secrets) can be reconciled in the syntax if we allow the subject nominal to occupy positions both below and above the two possible attachment sites for the _as_-prepositional clause.

![image](https://github.com/user-attachments/assets/53b15a31-f23e-481e-b3af-faad15a2c056)

(I'm not writing my dissertation on syntax but "I can draw a sentence as a tree" is a fun party trick)

**Reply by teunbrand** [reply]
> Wow I swear this is my first time hearing "workword" and "(independent/dependent) nameword"!

Yeah they're not valid terms in English but this is how I was thought about word identity/sentence structure in Dutch.

I have no idea what your tree means, but does this allow some sort of analysis of a corpus of text? Can you then say writer A prefers sentence structure ABC more than XYZ than writer B? Could this shed light on some plagiarism claims? E.g. in music it was kinda suspected that Mozart didn't finish the Requiem piece before his death and it was finished by someone else.

**Reply by EvaMaeRey** [reply]
This might have devolved more than opinionated package writers debating best practices! #53  Exciting to have a real linguist w/ syntax tree abilities in our midst!  The tree looks impressive! 

workword for verb is kind of poetic. I think I'd use verb/workword as the analogy for Stats/stat_*s if I were to return to the grammar guide analogy.  

In an early-ish talk on writing Stats to create new layers:
![image](https://github.com/user-attachments/assets/b2fce309-abdc-4a57-b9f6-4818bf919876)

engine -> mechanical work

probably wouldn't use the term 'statistical geometries' anymore - trying to say 'layer defined by new Stat and existing Geom'.  

I think that's my motivation Stats.  In some cases layers exist already to do a lot of computational work but sometimes you have to wrangle yourself - like @teunbrand formerly (and myself).  

> Before I knew how to use stats, I was always just wrangling data myself before handing the data off to some graphical primitive geom.

But I think people are actually pretty capable of conceptually understanding Stats and their value and following recipe pretty early on in their ggplot2 and even R journeys, and can construct useful layers for future even if they are not able to on-the-fly mix and match (I wonder at my own mix and match capability).

**Reply by yjunechoe** [reply]
> I have no idea what your tree means, but does this allow some sort of analysis of a corpus of text?

If you're willing to let go of some theoretical purity, then yeah! That's how "old school" NLP research started with, e.g., the [Penn Treebank](https://catalog.ldc.upenn.edu/docs/LDC95T7/cl93.html) project. Those models are useful for research but not super viable as products: they're very difficult to scale up (requires lots of heavily-annotated training data) and not particularly suitable for the more "creative" uses of language in the wild that intentionally flout rules of grammar (like internet language and even some literary genres).

> But I think people are actually pretty capable of conceptually understanding Stats and their value

I think so too!

**Reply by EvaMaeRey** [reply]
Federalist papers, published anonymously,  is a nice language who-done-it.  Hamilton, before his death apparently claimed some that later Madison claimed to have written.  Analyses conclude Madison probably did at least most of the work on disputed papers.  

federalist_network_with_node_features %>%  
  ggraph::ggraph(layout = "fr") +  
  ggraph::geom_edge_link(  
    aes(alpha = correlation)  
    ) +  
  ggraph::geom_node_point(  
    aes(color = author),  
    size = 6  
    ) +  
  theme_void() +  
  ggforce::geom_mark_hull(
    aes(x = x,
        y = y,
        fill = author)
    )

![image](https://github.com/user-attachments/assets/4dff776e-2300-4ceb-a8d8-0f4f78357789)
